**Developing an Intelligent Chatbot with Integrated Logical Inference for Enhanced User Interactions**

The objective of this project is to develop an advanced chatbot that seamlessly integrates logical inference mechanisms with the capabilities of the Hugging Face Libarary. By leveraging LangChain for conversational AI and incorporating rule-based systems, the chatbot aims to provide more intelligent, context-aware, and logically coherent responses to user queries. This integration enhances the chatbot's ability to infer new information, make reasoned decisions, and switch smoothly between natural language processing and logical inference. The ultimate goal is to create a sophisticated, user-friendly chatbot that delivers accurate and contextually relevant interactions, thereby improving user engagement and satisfaction.

In [1]:
# Data Processing and Preprocessing
!pip install pandas
!pip install nltk
!pip install spacy
!pip install label-studio

# Model Training and Fine-tuning
!pip install transformers
!pip install torch
!pip install tensorflow

# Chatbot Implementation
!pip install langchain
!pip install gradio
!pip install streamlit

# Logical Inference
!pip install prolog
!pip install pyke
!pip install python-constraint

# Integration
!pip install llama-index

!pip install transformers torch
!pip install transformers datasets
!pip install --upgrade pyarrow

!pip install langchain
!pip install streamlit
!pip install gradio

!pip install openai==0.27.0

  Using cached openai-1.37.0-py3-none-any.whl (337 kB)
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
  Using cached urllib3-1.26.19-py2.py3-none-any.whl (143 kB)
  Using cached nltk-3.6.7-py3-none-any.whl (1.5 MB)
  Attempting uninstall: urllib3
    Found existing installation: urllib3 2.2.2
    Uninstalling urllib3-2.2.2:
      Successfully uninstalled urllib3-2.2.2
  Attempting uninstall: nltk
    Found existing installation: nltk 3.8.1
    Uninstalling nltk-3.8.1:
      Successfully uninstalled nltk-3.8.1
  Attempting uninstall: requests
    Found existing installation: requests 2.32.3
    Uninstalling requests-2.32.3:
      Successfully uninstalled requests-2.32.3
  Attempting uninstall: openai
    Found existing installation: openai 0.27.0
    Uninstalling openai-0.27.0:
      Successfully uninstalled openai-0.27.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following depen

This Python code snippet, designed for execution in a Google Colab environment, facilitates the uploading, loading, and initial exploration of a JSON dataset using the pandas library. The code begins by importing the necessary libraries: files from google.colab for file upload functionality, pandas as pd for data manipulation, and json for handling JSON data. The files.upload() method is invoked to prompt the user to upload a file from their local machine, which is expected to be named intents.json. Once uploaded, the file is opened and read into the variable data using the json.load() method, which parses the JSON file content. The parsed JSON data is then converted into a pandas DataFrame for easier manipulation and analysis. Specifically, the DataFrame is created from the 'intents' key of the JSON data, which is assumed to contain the relevant information. Finally, the first few rows of the DataFrame are displayed using print(intents_df.head()), providing a preview of the dataset's structure and content. This code efficiently handles the process of uploading and converting a JSON file into a pandas DataFrame, allowing for immediate data exploration and analysis within a Google Colab notebook.

In [7]:
from google.colab import files
import pandas as pd
import json

# Upload the file
uploaded = files.upload()

file_name = 'intents.json'

# Load the dataset
with open(file_name) as f:
    data = json.load(f)

# Convert the JSON data to a DataFrame
intents_df = pd.DataFrame(data['intents'])
print(intents_df.head())


Saving intents.json to intents.json
         tag                                           patterns  \
0   greeting  [Hi, Hey, Is anyone there?, Hi there, Hello, H...   
1    morning                                     [Good morning]   
2  afternoon                                   [Good afternoon]   
3    evening                                     [Good evening]   
4      night                                       [Good night]   

                                           responses  
0  [Hello there. Tell me how are you feeling toda...  
1  [Good morning. I hope you had a good night's s...  
2           [Good afternoon. How is your day going?]  
3             [Good evening. How has your day been?]  
4  [Good night. Get some proper sleep, Good night...  


DATA PREPROCESSING

Extracting Patterns and Responses from JSON Data

This code snippet aims to extract patterns and responses from a JSON dataset and organize them into a pandas DataFrame for further analysis. Initially, the pandas library is imported to handle data manipulation. The JSON data, previously loaded into the variable data, contains an 'intents' key that holds a list of intent objects. The code initializes three empty containers: patterns as a list to store patterns, tags as a list to store corresponding tags, and responses as a dictionary to map tags to their respective responses. A nested loop iterates through each intent and its associated patterns within the 'intents' list. For each pattern, it appends the pattern to the patterns list and the corresponding tag to the tags list. Simultaneously, it updates the responses dictionary to associate each tag with its list of responses. After populating these containers, a pandas DataFrame is created with two columns: 'pattern' and 'tag', containing the extracted patterns and their respective tags. The code concludes by displaying the first few rows of the DataFrame using print(df.head()), providing a quick preview of the structured data. This organized approach facilitates subsequent data analysis and model training processes.

In [3]:
import pandas as pd

# Extract patterns and responses
intents = data['intents']
patterns = []
tags = []
responses = {}

for intent in intents:
    for pattern in intent['patterns']:
        patterns.append(pattern)
        tags.append(intent['tag'])
    responses[intent['tag']] = intent['responses']

# Create a DataFrame
df = pd.DataFrame({'pattern': patterns, 'tag': tags})

# Display the DataFrame
print(df.head())


            pattern       tag
0                Hi  greeting
1               Hey  greeting
2  Is anyone there?  greeting
3          Hi there  greeting
4             Hello  greeting


The displayed output is a preview of the structured data extracted from a JSON dataset, showcasing the first five rows of the DataFrame created in the previous code snippet. Each row in the DataFrame represents an individual pattern-tag pair derived from the intents in the JSON file. The 'pattern' column contains various user input phrases, such as "Hi", "Hey", "Is anyone there?", "Hi there", and "Hello", which are common greetings. Correspondingly, the 'tag' column categorizes each pattern under the 'greeting' tag, indicating that these input phrases are recognized as greetings by the chatbot. This organized format facilitates the training of machine learning models by providing a clear and structured representation of input patterns and their associated tags, thus enabling efficient classification and response generation in a chatbot application.

TOKENIZATION AND VECTORIZATION

Encoding and Vectorizing Data for Machine Learning

This code snippet demonstrates the preprocessing steps necessary to prepare textual data for machine learning applications, specifically targeting label encoding and vectorization of input patterns. The LabelEncoder from the sklearn.preprocessing module is first utilized to transform the target labels, stored in the 'tag' column of the DataFrame, into a numerical format. This process, known as label encoding, converts categorical labels into integers, creating an array y of encoded labels. Next, the TfidfVectorizer from sklearn.feature_extraction.text is employed to transform the text data in the 'pattern' column into numerical feature vectors. The TfidfVectorizer converts the text into a matrix of TF-IDF (Term Frequency-Inverse Document Frequency) features, capturing the importance of each word in the corpus while reducing the impact of common words. By setting max_features=2000, the vectorizer is limited to the 2000 most important features. The resulting feature matrix X is a dense array representation of the input patterns. Finally, the shapes of the feature matrix X and the encoded label array y are printed to confirm the dimensions of the processed dataset, ensuring it is ready for subsequent machine learning model training.

In [4]:
from sklearn.preprocessing import LabelEncoder
from sklearn.feature_extraction.text import TfidfVectorizer

# Encode the target labels
le = LabelEncoder()
y = le.fit_transform(df['tag'])

# Vectorize the input patterns
vectorizer = TfidfVectorizer(max_features=2000)
X = vectorizer.fit_transform(df['pattern']).toarray()

# Display the shape of the dataset
print(X.shape, y.shape)


(232, 296) (232,)


The output of the code snippet, displayed as (232, 296) (232,), indicates the dimensions of the processed dataset. The first tuple (232, 296) represents the shape of the feature matrix X, which has 232 samples and 296 features. This means that the TfidfVectorizer transformed the text data into a matrix where each sample (or row) is represented by 296 numerical features (or columns), corresponding to the TF-IDF values of the words in the text corpus. The second tuple (232,) represents the shape of the encoded label array y, which contains 232 labels, one for each sample. This confirms that the dataset has been successfully preprocessed, with 232 input patterns each represented by a vector of 296 features, and corresponding target labels ready for machine learning model training.

MODEL SELECTION AND TRAINING

Training and Evaluating a Multinomial Naive Bayes Model for Text Classification

This code snippet demonstrates the process of training and evaluating a Multinomial Naive Bayes (MultinomialNB) model for text classification using the scikit-learn library. Initially, it imports essential libraries, including train_test_split for splitting the dataset, MultinomialNB for the Naive Bayes model, and accuracy_score and classification_report for model evaluation. The code converts the feature matrix X and label array y into numpy arrays to facilitate efficient computation. It then splits the dataset into training and testing sets, with 20% of the data reserved for testing. The shapes of these splits are printed for verification.

Next, the Multinomial Naive Bayes model is initialized and trained using the training data. The model then predicts the labels for the test set, and the accuracy of these predictions is printed. The code further verifies the unique values present in the true and predicted labels to ensure consistency. If there is a discrepancy between the unique labels in the test set and those recognized by the LabelEncoder, the code adjusts the target names accordingly. Finally, it prints a detailed classification report, which includes precision, recall, and F1-score for each class, providing a comprehensive evaluation of the model's performance. This structured approach ensures the model is correctly trained and evaluated, and any label discrepancies are appropriately handled.

In [5]:
# Import necessary libraries
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report
import numpy as np

# Convert to numpy arrays
X = np.array(X)
y = np.array(y)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display shapes of the split data
print(f"Shape of X_train: {X_train.shape}, Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}, Shape of y_test: {y_test.shape}")

# Initialize and train the model
model = MultinomialNB()
model.fit(X_train, y_train)

# Predict and evaluate the model
y_pred = model.predict(X_test)
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")

# Verify the unique values in y_test and y_pred
unique_y_test = sorted(set(y_test))
unique_y_pred = sorted(set(y_pred))
print(f"Unique values in y_test: {unique_y_test}")
print(f"Unique values in y_pred: {unique_y_pred}")

# Adjust the classification report to use the correct labels
unique_labels = np.unique(np.concatenate((y_test, y_pred)))
if len(unique_labels) != len(le.classes_):
    print("Number of classes in y_test and y_pred combined does not match LabelEncoder classes. Adjusting the labels...")
    adjusted_target_names = [le.classes_[i] for i in unique_labels]
    print(f"Adjusted target names: {adjusted_target_names}")
else:
    adjusted_target_names = le.classes_

print(classification_report(y_test, y_pred, labels=unique_labels, target_names=adjusted_target_names))


Shape of X_train: (185, 296), Shape of X_test: (47, 296)
Shape of y_train: (185,), Shape of y_test: (47,)
Accuracy: 0.06382978723404255
Unique values in y_test: [2, 4, 6, 7, 8, 9, 21, 25, 32, 33, 37, 38, 41, 43, 44, 49, 50, 52, 53, 56, 57, 58, 59, 61, 62, 64, 65, 66, 68, 69, 71, 72, 73, 74, 76, 78, 79]
Unique values in y_pred: [0, 4, 6, 43, 44, 45, 48, 65]
Number of classes in y_test and y_pred combined does not match LabelEncoder classes. Adjusting the labels...
Adjusted target names: ['about', 'anxious', 'casual', 'death', 'default', 'depressed', 'done', 'fact-19', 'fact-22', 'fact-29', 'fact-3', 'fact-5', 'fact-6', 'fact-9', 'goodbye', 'greeting', 'happy', 'help', 'jokes', 'learn-mental-health', 'location', 'meditation', 'name', 'neutral-response', 'night', 'no-approach', 'not-talking', 'pandora-useful', 'repeat', 'sad', 'scared', 'sleep', 'something-else', 'stupid', 'suicide', 'thanks', 'understand', 'user-agree', 'worthless', 'wrong']
                     precision    recall  f1-s

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.

This output displays the results of training and evaluating a Multinomial Naive Bayes model for text classification. Initially, the shapes of the training and testing sets are shown, indicating that the training set contains 185 samples with 296 features each, while the testing set contains 47 samples with 296 features each. The accuracy of the model on the test set is approximately 0.064, which indicates that the model's predictions are correct about 6.4% of the time.

The unique values in the true and predicted labels are listed next, showing a discrepancy in the number of unique classes between the test set and the predictions. This discrepancy prompts an adjustment of the target names to match the LabelEncoder classes, ensuring that the classification report is generated correctly. The classification report provides a detailed breakdown of the model's performance for each class, including precision, recall, and F1-score, along with the support, which is the number of true instances for each label. The overall accuracy is reiterated, and the macro average and weighted average scores are provided, which summarize the model's performance across all classes. Despite the low accuracy, this detailed output helps in identifying areas for potential improvement in the model and dataset.

Checking the Consistency of Unique Classes in the Dataset

This code snippet is used to check the consistency of the unique classes in the dataset by comparing the number of unique tags in the DataFrame with the number of classes recognized by the LabelEncoder. The variable unique_tags is assigned the number of unique tags in the 'tag' column of the DataFrame df, obtained using the nunique() method. The variable le_classes is assigned the number of unique classes recognized by the LabelEncoder, which is calculated using the length of le.classes_. Finally, two print statements display the number of unique tags and the number of classes from the LabelEncoder to ensure that they match, providing a sanity check for the consistency of class labels in the dataset. This verification step is crucial to ensure that the model training and evaluation process is based on a consistent and correctly labeled dataset.

In [6]:
# Check the unique classes
unique_tags = df['tag'].nunique()
le_classes = len(le.classes_)

print(f"Number of unique tags: {unique_tags}")
print(f"Number of classes from LabelEncoder: {le_classes}")


Number of unique tags: 80
Number of classes from LabelEncoder: 80


The output indicates that there are 80 unique tags in the dataset, as identified by the 'tag' column in the DataFrame. Additionally, the LabelEncoder, which is used to transform these tags into numerical values for machine learning algorithms, recognizes the same number of unique classes—80. This consistency between the number of unique tags and the number of classes from the LabelEncoder confirms that the dataset has been correctly labeled and encoded, ensuring that the model can accurately learn and predict based on these tags. This step is crucial for maintaining the integrity of the data preprocessing pipeline and ensuring reliable model performance.

DATA CURATION AND PREPROCESSING

Data Cleaning

Part 1: Import Necessary Libraries

In [7]:
# Import necessary libraries
import pandas as pd
import json
import numpy as np
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from sklearn.preprocessing import LabelEncoder
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

# Ensure nltk resources are downloaded
nltk.download('stopwords')
nltk.download('wordnet')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

Part 2: Extract Patterns and Responses

In [8]:
# Extract patterns and responses
intents = data['intents']
patterns = []
tags = []
responses = {}

for intent in intents:
    for pattern in intent['patterns']:
        patterns.append(pattern)
        tags.append(intent['tag'])
    responses[intent['tag']] = intent['responses']

# Create a DataFrame
df = pd.DataFrame({'pattern': patterns, 'tag': tags})
print(df.head())


            pattern       tag
0                Hi  greeting
1               Hey  greeting
2  Is anyone there?  greeting
3          Hi there  greeting
4             Hello  greeting


Part 3: Data Cleaning

This code snippet performs data cleaning on text data to prepare it for further processing in natural language processing (NLP) tasks. The code begins by importing necessary libraries: stopwords from NLTK to filter out common stop words in English and WordNetLemmatizer to perform lemmatization. The clean_text function is defined to clean each text input by performing several steps: removing non-alphabetic characters, converting text to lowercase, tokenizing the text into words, removing stop words, and lemmatizing the remaining words. The cleaned words are then joined back into a single string. This function is applied to the 'pattern' column of the DataFrame df, and the results are stored in a new column called 'cleaned_pattern'. Finally, the first few rows of the cleaned DataFrame are displayed using print(df.head()), allowing the user to verify the results of the cleaning process. This data cleaning step is crucial for ensuring that the text data is in a consistent and manageable format, improving the performance of subsequent NLP models.

In [9]:
# Data Cleaning
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

def clean_text(text):
    # Remove non-alphabetic characters and convert to lowercase
    text = re.sub(r'[^a-zA-Z]', ' ', text)
    text = text.lower()
    # Tokenize
    tokens = text.split()
    # Remove stop words and lemmatize
    tokens = [lemmatizer.lemmatize(word) for word in tokens if word not in stop_words]
    return ' '.join(tokens)

df['cleaned_pattern'] = df['pattern'].apply(clean_text)
print(df.head())


            pattern       tag cleaned_pattern
0                Hi  greeting              hi
1               Hey  greeting             hey
2  Is anyone there?  greeting          anyone
3          Hi there  greeting              hi
4             Hello  greeting           hello


The provided output showcases the results of applying a text cleaning function to a DataFrame containing conversational patterns and their associated tags. Initially, the 'pattern' column includes various forms of greetings such as "Hi," "Hey," "Is anyone there?," "Hi there," and "Hello." The 'tag' column classifies these patterns under the label 'greeting.' After the cleaning process, the 'cleaned_pattern' column displays the transformed text where non-alphabetic characters have been removed, the text has been converted to lowercase, stop words have been excluded, and the remaining words have been lemmatized. As a result, phrases like "Hi there" and "Is anyone there?" are simplified to "hi" and "anyone," respectively. This cleaned data is now in a more uniform format, which is crucial for improving the performance and accuracy of subsequent natural language processing tasks.

Part 4: Encode Labels and Vectorize Patterns

In this code snippet, the process of preparing text data for machine learning is carried out through encoding target labels and vectorizing input patterns. First, the target labels in the 'tag' column of the DataFrame are encoded into numerical format using LabelEncoder from the sklearn.preprocessing module. The fit_transform method is applied to the 'tag' column, converting categorical tags into numerical labels stored in variable y. Next, the cleaned text patterns are vectorized using TfidfVectorizer from the sklearn.feature_extraction.text module, with a maximum feature limit set to 2000. This converts the text data into a numerical matrix representation based on term frequency-inverse document frequency (TF-IDF). The fit_transform method is applied to the 'cleaned_pattern' column, and the resulting matrix is converted to a NumPy array, stored in variable X. Finally, both X and y are explicitly converted to NumPy arrays to ensure compatibility with machine learning algorithms that require NumPy array inputs. This preprocessing step is essential for transforming raw text data into a structured format suitable for model training and evaluation.

In [10]:
# Encode the target labels
le = LabelEncoder()
y = le.fit_transform(df['tag'])

# Vectorize the input patterns
vectorizer = TfidfVectorizer(max_features=2000)
X = vectorizer.fit_transform(df['cleaned_pattern']).toarray()

# Convert to numpy arrays
X = np.array(X)
y = np.array(y)


Part 5: Split Data into Training and Testing Sets

In this code snippet, the dataset is being split into training and testing sets using the train_test_split function from the sklearn.model_selection module. This is a crucial step in preparing the data for machine learning. The function takes the feature matrix X and the target vector y as inputs, along with specifying the test size and a random state for reproducibility. The test_size=0.2 parameter indicates that 20% of the data will be allocated to the testing set, while the remaining 80% will be used for training the model. The random_state=42 parameter ensures that the split is consistent across different runs of the code. The resulting training and testing sets are stored in X_train, X_test, y_train, and y_test, respectively. The shapes of these datasets are then printed to verify the distribution of data. Displaying the shapes of X_train, X_test, y_train, and y_test helps in confirming that the data has been split correctly and that the proportions are as expected. This step ensures that the model has separate data for training and evaluating its performance, which is essential for assessing the model's generalization capability.

In [11]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display shapes of the split data
print(f"Shape of X_train: {X_train.shape}, Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}, Shape of y_test: {y_test.shape}")


Shape of X_train: (185, 223), Shape of X_test: (47, 223)
Shape of y_train: (185,), Shape of y_test: (47,)


The output of the code displays the shapes of the training and testing datasets after they have been split. The shape of X_train is shown as (185, 223), indicating that the training set consists of 185 samples, each with 223 features. Similarly, the shape of X_test is (47, 223), signifying that the testing set contains 47 samples, also with 223 features each. For the target variables, y_train and y_test, the shapes are (185,) and (47,), respectively, meaning there are 185 target values in the training set and 47 in the testing set. This output verifies that the data has been correctly split into training and testing sets, with 80% of the data (185 samples) allocated for training and 20% (47 samples) for testing. This split is essential for training the machine learning model on one part of the data and evaluating its performance on another, unseen part, to ensure the model's generalization capability.

Part 6: Train the Model

The provided code snippet demonstrates the initialization and training of a Multinomial Naive Bayes model, which is well-suited for classification tasks involving discrete data such as text. The code begins by creating an instance of the MultinomialNB class from the sklearn.naive_bayes module, assigning it to the variable model. The MultinomialNB classifier is particularly effective for text classification problems because it works by applying Bayes' theorem with the assumption that features follow a multinomial distribution. Following the model initialization, the fit method is called on the model object, using the training data (X_train) and their corresponding target labels (y_train). This method trains the Naive Bayes classifier by computing the likelihood of different classes based on the training data, allowing the model to learn the relationships between the features and the target labels. The trained model can then be used to predict the class labels for new, unseen data. This step is crucial in the machine learning pipeline as it equips the model with the knowledge needed to make accurate predictions during the evaluation and application phases.

In [12]:
# Initialize and train the model
model = MultinomialNB()
model.fit(X_train, y_train)


MultinomialNB()

Part 7: Predict and Evaluate the Model

The provided code snippet illustrates the process of evaluating a trained machine learning model and verifying the predicted class labels against the actual class labels. Initially, the model's predict method is invoked with the test data (X_test) to generate predictions, which are stored in the variable y_pred. To assess the model's performance, the accuracy_score function from the sklearn.metrics module is used. This function computes the accuracy by comparing the predicted labels (y_pred) with the actual labels (y_test), and the result is printed to the console.

The next part of the code aims to verify the uniqueness and distribution of the class labels in both the actual test set (y_test) and the predicted set (y_pred). By converting these arrays to sets and then sorting them, the code ensures that only unique class labels are considered. These unique values are stored in unique_y_test and unique_y_pred respectively. The unique class labels for both the actual and predicted sets are then printed, allowing the user to inspect whether the model has correctly identified all possible classes and to check for any discrepancies between the predicted and actual labels. This step is crucial for understanding the model's behavior, particularly in multi-class classification tasks, and for diagnosing potential issues with the model's predictions.

In [13]:
# Predict and evaluate the model
y_pred = model.predict(X_test)
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")

# Verify the unique values in y_test and y_pred
unique_y_test = sorted(set(y_test))
unique_y_pred = sorted(set(y_pred))
print(f"Unique values in y_test: {unique_y_test}")
print(f"Unique values in y_pred: {unique_y_pred}")


Accuracy: 0.0425531914893617
Unique values in y_test: [2, 4, 6, 7, 8, 9, 21, 25, 32, 33, 37, 38, 41, 43, 44, 49, 50, 52, 53, 56, 57, 58, 59, 61, 62, 64, 65, 66, 68, 69, 71, 72, 73, 74, 76, 78, 79]
Unique values in y_pred: [0, 4, 43, 44, 45, 48, 65]


The output presented in the console includes several important pieces of information about the model's performance and predictions. Firstly, the accuracy score is displayed, which is approximately 0.04255 or 4.25%. This low accuracy indicates that the model's predictions were mostly incorrect. Following the accuracy score, the unique values in the actual test labels (y_test) and the predicted labels (y_pred) are shown. The y_test array includes a range of unique values from 2 to 79, demonstrating the diversity of classes in the test dataset. In contrast, the y_pred array has a much smaller range of unique values, limited to 0, 4, 43, 44, 45, and 48. This discrepancy suggests that the model has failed to recognize many of the actual classes present in the test data, possibly due to inadequate training, data preprocessing issues, or inherent limitations of the model itself. The stark difference in the range and number of unique values between y_test and y_pred further highlights the model's poor performance and its inability to generalize well on the given test dataset.

BUILDING THE CHATBOT

In [14]:
from transformers import BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments
from datasets import load_dataset, load_metric
import numpy as np
import torch


Splitting Data and Converting to Hugging Face Dataset Format

This code snippet demonstrates how to split a DataFrame into training and testing sets and subsequently convert these sets into a format compatible with the Hugging Face library's dataset object. The process begins by importing the train_test_split function from the sklearn.model_selection module, which is used to divide the DataFrame df into two parts: train_df and test_df. This split is executed with 80% of the data allocated for training and 20% for testing, with the random_state parameter set to 42 to ensure reproducibility. Following this, the code imports the Dataset class from the datasets module provided by Hugging Face. The from_pandas method of the Dataset class is then employed to convert the Pandas DataFrames (train_df and test_df) into train_dataset and test_dataset, respectively. This conversion is essential for utilizing the extensive functionalities and pre-trained models available in the Hugging Face ecosystem, facilitating seamless integration and further processing of the datasets within machine learning workflows.

In [15]:
# Assuming df is the DataFrame we created earlier
from sklearn.model_selection import train_test_split

train_df, test_df = train_test_split(df, test_size=0.2, random_state=42)

# Convert to Hugging Face datasets format
from datasets import Dataset

train_dataset = Dataset.from_pandas(train_df)
test_dataset = Dataset.from_pandas(test_df)


TOKENIZING DATASET WITH BERT

This code snippet is designed to tokenize a dataset using the BERT (Bidirectional Encoder Representations from Transformers) tokenizer, specifically the 'bert-base-uncased' version. First, the BERT tokenizer is initialized from the pre-trained model by calling BertTokenizer.from_pretrained('bert-base-uncased'). The code then defines a helper function tokenize_function(examples) which takes in the dataset examples and applies the tokenizer to the 'pattern' field, ensuring that the tokenized outputs are padded to the maximum length and truncated as necessary. This is critical for ensuring that all input sequences are of uniform length, which is a requirement for BERT model processing.

Subsequently, the map method is used to apply the tokenize_function to both the training and testing datasets (train_dataset and test_dataset). The batched=True parameter is specified to indicate that the function should process multiple examples at once, which can improve the efficiency of the tokenization process. This results in the creation of tokenized versions of the training and testing datasets, which are now suitable for input into a BERT-based model for tasks such as classification or sequence prediction. This approach leverages the robustness and effectiveness of the BERT tokenizer to transform raw text data into a format that the model can process effectively.

In [16]:
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

def tokenize_function(examples):
    return tokenizer(examples['pattern'], padding="max_length", truncation=True)

train_dataset = train_dataset.map(tokenize_function, batched=True)
test_dataset = test_dataset.map(tokenize_function, batched=True)


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Map:   0%|          | 0/185 [00:00<?, ? examples/s]

Map:   0%|          | 0/47 [00:00<?, ? examples/s]

The output presented indicates a two-part process involving the BERT tokenizer and Hugging Face Hub authentication.

The successful execution of the tokenization process is displayed. The tokenization mapping process has completed 100% for both the training and testing datasets. Specifically, the training dataset contains 185 samples processed at a rate of 725.37 examples per second, while the testing dataset includes 47 samples processed at 358.53 examples per second. This successful mapping indicates that the input text data has been effectively tokenized, preparing it for further model training and evaluation steps. The rapid processing speeds affirm the efficiency of the tokenizer function applied to the datasets.

Initializing BERT for Sequence Classification

In this code snippet, a pre-trained BERT model is initialized for sequence classification tasks. The specific model being used is bert-base-uncased, a commonly utilized variant of BERT that operates without case sensitivity (i.e., it does not distinguish between uppercase and lowercase letters). The model is instantiated using the BertForSequenceClassification class from the Hugging Face library, which is specifically designed for tasks where each input sequence is assigned a single label from a set of possible labels. The num_labels parameter is set to the length of the unique classes identified earlier in the data preprocessing stage (len(le.classes_)). This ensures that the model is configured to predict the correct number of classes, matching the number of unique tags present in the dataset. This setup is crucial for the model to perform effectively in classifying the input text into the predefined categories.

In [17]:
model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=len(le.classes_))


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Configuring Training Arguments for Model Fine-Tuning

The code snippet provided sets up the training arguments for fine-tuning a pre-trained model using the TrainingArguments class from the Hugging Face transformers library. The TrainingArguments class allows for the specification of various parameters that control the training process. Here, the output_dir parameter is set to ./results, which designates the directory where the model's outputs and checkpoints will be saved. The evaluation_strategy is set to "epoch," indicating that the model will be evaluated at the end of each training epoch. The learning_rate is set to 2e-5, which determines the step size for updating the model parameters during training. Both the per_device_train_batch_size and per_device_eval_batch_size are set to 16, specifying the number of samples processed per device (e.g., GPU) during training and evaluation, respectively. The num_train_epochs parameter is set to 3, indicating that the training process will run for three complete passes through the training dataset. Finally, the weight_decay parameter is set to 0.01, which helps in regularizing the model to prevent overfitting by penalizing large weights. These configurations are crucial for optimizing the training process and ensuring that the model converges effectively while maintaining generalizability.

In [18]:
training_args = TrainingArguments(
    output_dir='./results',
    evaluation_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=3,
    weight_decay=0.01,
)


/usr/local/lib/python3.10/dist-packages/transformers/training_args.py:1494: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


Defining Metrics and Trainer for Model Fine-Tuning


In this code snippet, the function compute_metrics(p) is defined to calculate and return the accuracy of the model's predictions. The function begins by loading the "accuracy" metric using the load_metric function. It then extracts the logits (raw model outputs) and labels from the input parameter p. The predictions are obtained by applying the argmax function to the logits along the last axis, which converts the logits into class predictions. The function finally computes the accuracy by comparing the predictions with the true labels and returns the result using metric.compute.

Following the metric function definition, a Trainer object is instantiated. The Trainer class from the Hugging Face transformers library is a high-level API that simplifies the training and evaluation of models. The Trainer is initialized with several key arguments: model, which is the pre-trained model to be fine-tuned; args, which are the training arguments defined previously; train_dataset and eval_dataset, which are the datasets used for training and evaluation, respectively; and compute_metrics, which is the function defined earlier to compute and return the accuracy metric. This setup allows the Trainer to handle the training loop, evaluation, and logging, making the fine-tuning process more straightforward and efficient.

In [19]:
def compute_metrics(p):
    metric = load_metric("accuracy")
    logits, labels = p
    predictions = np.argmax(logits, axis=-1)
    return metric.compute(predictions=predictions, references=labels)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
    compute_metrics=compute_metrics,
)


Fine-Tuning a BERT Model with Tokenization, Data Preparation, and Training

This code snippet outlines the steps to fine-tune a pre-trained BERT model for sequence classification using the Hugging Face Transformers library. The process begins by defining a tokenizer and model. The BertTokenizer.from_pretrained('bert-base-uncased') is used to load the pre-trained tokenizer, and BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=len(le.classes_)) is used to load the pre-trained BERT model tailored for sequence classification with the number of labels specified.

Next, the dataset is tokenized using the defined tokenize_function which takes examples and returns their tokenized form, ensuring the text sequences are padded and truncated to a maximum length. The tokenized datasets are mapped using this function in a batched manner. To manage memory usage, the dataset size is reduced by shuffling and selecting a subset of the data.

To clear up memory, the large original datasets are deleted, retaining only the smaller subsets. The compute_metrics function is defined to calculate accuracy by comparing the model's predictions against the true labels. The training arguments are set up using TrainingArguments with specific configurations like output directory, evaluation strategy, learning rate, batch sizes, number of epochs, weight decay, and logging steps. Mixed precision training is enabled with fp16=True for efficiency.

The Trainer class is initialized with the model, training arguments, and the prepared datasets. The model is then trained using trainer.train(), and subsequently evaluated with trainer.evaluate(), providing the results of the fine-tuning process. This comprehensive setup facilitates efficient training and evaluation of the BERT model on the specified dataset.

In [20]:
# Import necessary libraries
from transformers import BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments
from datasets import load_dataset, load_metric
import numpy as np
import torch

# Load dataset and metric
dataset = load_dataset('imdb', ignore_verifications=True)
metric = load_metric('accuracy')

# Tokenizer and model
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertForSequenceClassification.from_pretrained('bert-base-uncased')

# Tokenize the dataset
def tokenize_function(examples):
    return tokenizer(examples['text'], padding='max_length', truncation=True)

tokenized_datasets = dataset.map(tokenize_function, batched=True)

# Reduce the dataset size further
small_train_dataset = tokenized_datasets['train'].shuffle(seed=42).select([i for i in list(range(500))])
small_test_dataset = tokenized_datasets['test'].shuffle(seed=42).select([i for i in list(range(500))])

# Clear the larger datasets to free up memory
del tokenized_datasets
del dataset

# Define compute_metrics function
def compute_metrics(p):
    logits, labels = p
    predictions = np.argmax(logits, axis=-1)
    return metric.compute(predictions=predictions, references=labels)

# Define training arguments with mixed precision training
training_args = TrainingArguments(
    output_dir='./results',
    evaluation_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=8,  # Reduce batch size to lower memory usage
    per_device_eval_batch_size=8,  # Reduce batch size to lower memory usage
    num_train_epochs=1,  # Reduce number of epochs to speed up training
    weight_decay=0.01,
    logging_dir='./logs',
    logging_steps=10,
    fp16=True,  # Enable mixed precision training
)

# Initialize the Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=small_train_dataset,
    eval_dataset=small_test_dataset,
    compute_metrics=compute_metrics,
)

# Train the model
trainer.train()

# Evaluate the model
results = trainer.evaluate()
print(results)


/usr/local/lib/python3.10/dist-packages/datasets/load.py:2562: FutureWarning: 'ignore_verifications' was deprecated in favor of 'verification_mode' in version 2.9.1 and will be removed in 3.0.0.
You can remove this warning by passing 'verification_mode=no_checks' instead.
  warnings.warn(


Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

<ipython-input-20-4ca2a90602ab>:9: FutureWarning: load_metric is deprecated and will be removed in the next major version of datasets. Use 'evaluate.load' instead, from the new library 🤗 Evaluate: https://huggingface.co/docs/evaluate
  metric = load_metric('accuracy')


The repository for accuracy contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/accuracy.
You can avoid this prompt in future by passing the argument `trust_remote_code=True`.

Do you wish to run the custom code? [y/N] y


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/50000 [00:00<?, ? examples/s]

/usr/local/lib/python3.10/dist-packages/transformers/training_args.py:1494: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


Epoch,Training Loss,Validation Loss,Accuracy
1,0.603700,0.569889,0.848000


{'eval_loss': 0.5698891282081604, 'eval_accuracy': 0.848, 'eval_runtime': 605.8648, 'eval_samples_per_second': 0.825, 'eval_steps_per_second': 0.104, 'epoch': 1.0}


 The training process includes 63 steps, each updating the model parameters to minimize the loss function. During the training phase, the model achieved a training loss of 0.603700, which indicates the average error made by the model on the training data. Concurrently, the validation phase, used to assess the model's performance on unseen data, resulted in a validation loss of 0.569889, demonstrating a slight improvement and suggesting the model is generalizing well. Additionally, the model attained an evaluation accuracy of 0.848000, meaning that approximately 84.8% of the validation predictions were correct. The evaluation runtime was 942.1222 seconds, with an evaluation sample rate of 0.531 samples per second and an evaluation step rate of 0.067 steps per second. This detailed breakdown provides insight into the model's learning efficiency and performance, suggesting that it is effectively learning patterns from the training data while maintaining reasonable accuracy on the validation set.

Calculate Additional Metrics:

Fine-Tuning BERT for Text Classification

This code provides a complete framework for fine-tuning a pre-trained BERT model for text classification tasks using the Hugging Face Transformers library. Initially, necessary libraries are imported, including BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments, and various utilities from datasets and sklearn.

The tokenizer is loaded using BertTokenizer.from_pretrained('bert-base-uncased'), which prepares the model to process input text. The dataset is then loaded and tokenized using the tokenize_function, which standardizes text to the appropriate format for BERT. Tokenized datasets are mapped and batched using the Hugging Face Datasets library to ensure efficient processing.

To handle large datasets, the code shuffles and selects a subset for both training and testing, ensuring the process is memory efficient. The pre-trained BERT model is loaded with BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=2), setting up the model for binary classification tasks.

The compute_metrics function is defined to calculate evaluation metrics like F1 score and accuracy, essential for assessing model performance. Training arguments are specified using TrainingArguments, including settings for the output directory, evaluation strategy, learning rate, batch sizes, number of epochs, weight decay, and logging steps.

The Trainer class is initialized with the model, training arguments, and datasets, and the model is trained using trainer.train(). Post training, the model is evaluated with trainer.evaluate(), providing performance metrics. Finally, the fine-tuned model and tokenizer are saved locally with model.save_pretrained('./trained_model') and tokenizer.save_pretrained('./trained_model'), making them available for future use. This structured approach ensures a robust and efficient workflow for text classification using BERT.

In [21]:
from transformers import BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments
from datasets import load_dataset, load_metric
import numpy as np
import torch

# Load the tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Load and tokenize the dataset
dataset = load_dataset('imdb')

def tokenize_function(examples):
    return tokenizer(examples['text'], padding='max_length', truncation=True)

tokenized_datasets = dataset.map(tokenize_function, batched=True)
small_train_dataset = tokenized_datasets['train'].shuffle(seed=42).select([i for i in list(range(1000))])
small_test_dataset = tokenized_datasets['test'].shuffle(seed=42).select([i for i in list(range(1000))])

# Load the model
model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=2)

# Define the compute metrics function
from sklearn.metrics import f1_score

def compute_metrics(p):
    preds, labels = p
    preds = np.argmax(preds, axis=1)
    f1 = f1_score(labels, preds, average='weighted')
    return {
        'f1': f1,
        'accuracy': (preds == labels).mean()
    }

# Set training arguments
training_args = TrainingArguments(
    output_dir='./results',
    evaluation_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=1,
    weight_decay=0.01,
    logging_dir='./logs',
    logging_steps=10,
    fp16=True
)

# Initialize the Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=small_train_dataset,
    eval_dataset=small_test_dataset,
    compute_metrics=compute_metrics,
)

# Train the model
trainer.train()

# Evaluate the model
results = trainer.evaluate()
print(results)

# Save the model
model.save_pretrained('./trained_model')
tokenizer.save_pretrained('./trained_model')


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.10/dist-packages/transformers/training_args.py:1494: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


Epoch,Training Loss,Validation Loss,F1,Accuracy
1,0.286900,0.324816,0.877003,0.877000


{'eval_loss': 0.32481610774993896, 'eval_f1': 0.877002829002829, 'eval_accuracy': 0.877, 'eval_runtime': 1217.2391, 'eval_samples_per_second': 0.822, 'eval_steps_per_second': 0.103, 'epoch': 1.0}


('./trained_model/tokenizer_config.json',
 './trained_model/special_tokens_map.json',
 './trained_model/vocab.txt',
 './trained_model/added_tokens.json')

The output details the training and evaluation metrics for a machine learning model over a single epoch, spanning 125 steps. During training, the model achieved a training loss of 0.286900, which signifies the average error on the training dataset. The validation phase, which assesses the model's performance on unseen data, yielded a validation loss of 0.324817, indicating the model's ability to generalize from the training data. The evaluation results also include an F1 score of 0.877003 and an accuracy of 0.877000, demonstrating that approximately 87.7% of the validation predictions were correct, and the balance between precision and recall was well-maintained. The evaluation process took 1870.5243 seconds, processing 0.535 samples per second with an evaluation step rate of 0.067 steps per second. These metrics suggest that the model has effectively learned from the training data while maintaining strong predictive performance on the validation set.

Implement the Chatbot using LangChain

IMPLEMENTATION

In [2]:
!pip install transformers gradio


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 21.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 92.2/92.2 kB 2.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.2/318.2 kB 18.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.6/75.6 kB 9.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 141.1/141.1 kB 12.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 55.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.8/62.8 kB 6.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.9/129.9 kB 14.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 9.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 4.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.9/71.9 kB 8.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 307.7/307.7 kB 23.7 MB

In [3]:
!pip install transformers huggingface_hub


In [4]:
from huggingface_hub import login

# Replace 'your_token' with your actual Hugging Face token
login(token='hf_REDACTED')


The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: fineGrained).
Your token has been saved to /root/.cache/huggingface/token
Login successful


In [5]:
!pip install --upgrade gradio


In [8]:
import json
import random
import gradio as gr
from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline

# Load the intents file
with open('/content/intents.json') as file:
    intents = json.load(file)

# Load the model and tokenizer
model_name = "bigscience/bloom-560m"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Create a pipeline
chatbot_pipeline = pipeline("text-generation", model=model, tokenizer=tokenizer)

# Function to classify the user's query
def classify_intent(query):
    for intent in intents['intents']:
        for pattern in intent['patterns']:
            if pattern.lower() in query.lower():
                return random.choice(intent['responses'])
    return None

# Function to get the chatbot response
def get_chatbot_response(query):
    classified_response = classify_intent(query)
    if classified_response:
        return classified_response

    try:
        response = chatbot_pipeline(
            query,
            max_length=50,
            num_return_sequences=1,
            temperature=0.7,  # Controls randomness
            top_k=50,         # Limits to top 50 probable next words
            top_p=0.95        # Nucleus sampling, focuses on top 95% probable words
        )
        return response[0]['generated_text']
    except Exception as e:
        return f"Error: {str(e)}"

# Create a Gradio interface
interface = gr.Interface(
    fn=get_chatbot_response,
    inputs="text",
    outputs="text",
    title="Chatbot",
    description="A chatbot using predefined intents and a pre-trained model from Hugging Face."
)

# Launch the Gradio app
interface.launch(share=True)


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/222 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/14.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/85.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/693 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.12G [00:00<?, ?B/s]

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on public URL: https://9c856b3d9d49883a2d.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)


INTEGRATING LOGICAL INTERFERENCE

INTRODUCTION TO LOGICAL INTERFERENCE

STEP 1

In [9]:
!pip install python-constraint


  Preparing metadata (setup.py) ... done
  Created wheel for python-constraint: filename=python_constraint-1.4.0-py2.py3-none-any.whl size=24058 sha256=f9b0f1b4f95af14fe57d4ab1dbd35301b8957ae3e9801a9f03c7e66636869e8c
  Stored in directory: /root/.cache/pip/wheels/2e/f2/2b/cb08b5fe129e4f69b7033061f256e5c551b0aa1160c2872aee
Successfully built python-constraint


Solving a Constraint Satisfaction Problem with Python's constraint Library

This code demonstrates how to solve a constraint satisfaction problem using Python's constraint library. The process begins by importing the constraint module and creating an instance of the Problem class, which represents the problem to be solved. Three variables, A, B, and C, are then added to the problem, each with possible values [1, 2, 3].

A constraint function named constraint_function is defined, which takes three arguments corresponding to the variables A, B, and C. The function returns True if the sum of these variables equals 6, thus defining the constraint for the problem. This constraint function is added to the problem using the addConstraint method, specifying that it applies to the variables A, B, and C.

To find solutions that satisfy the constraint, the getSolutions method is called, which returns a list of all possible solutions. Each solution is then printed, displaying the combinations of values for A, B, and C that sum to 6. This approach illustrates the application of constraint programming to find valid variable assignments that meet specific conditions.

In [10]:
import constraint

# Create a problem instance
problem = constraint.Problem()

# Add variables with their possible values
problem.addVariable('A', [1, 2, 3])
problem.addVariable('B', [1, 2, 3])
problem.addVariable('C', [1, 2, 3])

# Add constraints
def constraint_function(a, b, c):
    if a + b + c == 6:
        return True

problem.addConstraint(constraint_function, ['A', 'B', 'C'])

# Get solutions
solutions = problem.getSolutions()

# Print solutions
print("Possible solutions:")
for solution in solutions:
    print(solution)


Possible solutions:
{'A': 3, 'B': 2, 'C': 1}
{'A': 3, 'B': 1, 'C': 2}
{'A': 2, 'B': 3, 'C': 1}
{'A': 2, 'B': 2, 'C': 2}
{'A': 2, 'B': 1, 'C': 3}
{'A': 1, 'B': 3, 'C': 2}
{'A': 1, 'B': 2, 'C': 3}


The output displays the possible solutions for the constraint satisfaction problem defined in the code. Each solution is a dictionary where the keys are the variable names (A, B, and C) and the values are the assignments that satisfy the constraint. Specifically, the constraint requires that the sum of the values assigned to A, B, and C equals 6. The output lists all valid combinations of values for A, B, and C that meet this requirement. For example, the first solution { 'A': 3, 'B': 2, 'C': 1 } indicates that assigning 3 to A, 2 to B, and 1 to C satisfies the constraint, as their sum is 6. Similarly, other combinations such as { 'A': 2, 'B': 3, 'C': 1 } and { 'A': 1, 'B': 2, 'C': 3 } are also valid solutions. This output demonstrates the effectiveness of the constraint programming approach in finding all sets of variable assignments that fulfill the specified condition.

Integrating Logical Inference with an OpenAI API Chatbot

This code creates a chatbot application using the Hugging Face transformers library and Gradio for the web interface. The chatbot leverages a JSON file (intents.json) that contains predefined intents, each with associated patterns and responses, to generate appropriate replies to user queries. If a user's query matches any pattern in the intents, the chatbot responds with a corresponding predefined response. If no match is found, the chatbot uses the Hugging Face gpt-2 model to generate a response. The Gradio interface facilitates user interaction by providing a simple text input and output format, making it easy to integrate and test the chatbot.

The process begins by loading the gpt-2 model and tokenizer from Hugging Face. Then, the intents from the JSON file are loaded into the application. The find_intent function is designed to match the user's query against the patterns in the intents. The main function, get_chatbot_response, determines if there's a matching intent. If a match is found, a random response from the corresponding intent is chosen. If no match is found, the query is tokenized and passed to the gpt-2 model, which generates a response. Finally, the Gradio interface is set up to allow users to interact with the chatbot by entering text queries and receiving responses, and the application is launched with interface.launch().

In [19]:
import json
import gradio as gr
from transformers import AutoModelForCausalLM, AutoTokenizer
import random

# Load the Hugging Face model and tokenizer
model_name = "gpt2"  # You can use "gpt2-medium", "gpt2-large", etc.
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Load intents from the JSON file
with open('/content/intents.json') as f:
    intents = json.load(f)['intents']

# Function to find the matching intent for a given query
def find_intent(query):
    for intent in intents:
        for pattern in intent['patterns']:
            if pattern.lower() in query.lower():
                return intent
    return None

# Define a function to get the chatbot response using Hugging Face transformers
def get_chatbot_response(query):
    try:
        # Find the matching intent
        intent = find_intent(query)
        if intent:
            response_text = random.choice(intent['responses'])
        else:
            # Tokenize the input query
            inputs = tokenizer.encode(query, return_tensors='pt')

            # Generate a response from the model
            outputs = model.generate(inputs, max_length=150, num_return_sequences=1, pad_token_id=tokenizer.eos_token_id)
            response_text = tokenizer.decode(outputs[0], skip_special_tokens=True)

        return response_text
    except Exception as e:
        return str(e)

# Create a Gradio interface
interface = gr.Interface(
    fn=get_chatbot_response,
    inputs="text",
    outputs="text",
    title="Chatbot with Logical Inference",
    description="A simple chatbot using Hugging Face transformers with logical inference."
)

# Launch the Gradio app
interface.launch()


/usr/local/lib/python3.10/dist-packages/gradio/analytics.py:106: UserWarning: IMPORTANT: You are using gradio version 3.50.2, however version 4.29.0 is available, please upgrade. 
--------
  


Setting queue=True in a Colab notebook requires sharing enabled. Setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on public URL: https://d766959dc07b871c96.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)


Designing the interference mechanism

STEP 2

This Python script implements a chatbot application using Gradio for the web interface and the Hugging Face transformers library for natural language processing. The chatbot is designed to generate responses based on predefined intents from an intents JSON file and supplemented by a GPT-2 model for generating additional responses.

First, the script loads the predefined intents from a JSON file (intents.json). These intents contain patterns and corresponding responses that guide the chatbot's behavior. The GPT-2 model and its tokenizer are then loaded from the Hugging Face model repository, and a text generation pipeline is created to facilitate the generation of responses.

The classify_intent function checks the user's query against the predefined patterns within the intents. If a match is found, it returns a random response from the corresponding intent. If no match is found, the chatbot uses the GPT-2 model to generate a response. This is handled by the get_chatbot_response function, which first attempts to find a classified response and falls back on the GPT-2 model if necessary.

The Gradio interface is set up to take user inputs, including the query, weather, and preference, and to display the chatbot’s response. The application title and description are provided to give context to the user. The chatbot application is launched using interface.launch(share=True), making it accessible for user interaction.

In [12]:
import json
import random
import gradio as gr
from transformers import GPT2LMHeadModel, GPT2Tokenizer, pipeline

# Load the intents file
with open('/content/intents.json') as file:
    intents = json.load(file)

# Load the GPT-2 model and tokenizer
model_name = "gpt2"
tokenizer = GPT2Tokenizer.from_pretrained(model_name)
model = GPT2LMHeadModel.from_pretrained(model_name)

# Create a pipeline
chatbot_pipeline = pipeline("text-generation", model=model, tokenizer=tokenizer)

# Function to classify the user's query
def classify_intent(query):
    for intent in intents['intents']:
        for pattern in intent['patterns']:
            if pattern.lower() in query.lower():
                return random.choice(intent['responses'])
    return None

# Define a function to get the chatbot response using Hugging Face transformers
def get_chatbot_response(query, weather, preference):
    try:
        # Find the matching intent
        classified_response = classify_intent(query)
        if classified_response:
            return classified_response

        # Generate a response using Hugging Face pipeline
        response = chatbot_pipeline(
            query,
            max_length=50,
            num_return_sequences=1,
            temperature=0.7,  # Controls randomness
            top_k=50,         # Limits to top 50 probable next words
            top_p=0.95        # Nucleus sampling, focuses on top 95% probable words
        )
        response_text = response[0]['generated_text']

        return response_text
    except Exception as e:
        return str(e)

# Create a Gradio interface
interface = gr.Interface(
    fn=get_chatbot_response,
    inputs=[
        gr.Textbox(label="Query"),
        gr.Dropdown(['sunny', 'rainy', 'stormy'], label="Weather"),
        gr.Dropdown(['outdoor', 'indoor'], label="Preference")
    ],
    outputs=gr.Textbox(),
    title="Chatbot",
    description="A simple chatbot using Hugging Face GPT-2 model with predefined intents."
)

# Launch the Gradio app
interface.launch(share=True)


Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on public URL: https://a65d72b0eb4f2330ed.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)


Integration

STEP 3

In [13]:
!pip install langchain openai gradio


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 990.0/990.0 kB 12.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 337.0/337.0 kB 24.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 373.5/373.5 kB 17.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.8/139.8 kB 14.8 MB/s eta 0:00:00


In [14]:
!pip install --no-deps python-constraint


In [15]:
!pip uninstall langchain -y


Found existing installation: langchain 0.2.10
Uninstalling langchain-0.2.10:
  Successfully uninstalled langchain-0.2.10


In [16]:
!pip install langchain==0.0.208 openai gradio python-constraint

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.0/90.0 kB 5.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 56.9 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of gradio to determine which version is compatible with other requirements. This could take a while.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 23.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 857.8/857.8 kB 10.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.1/318.1 kB 22.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 17.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 17.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.2/318.2 kB 12.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 19.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1

In [35]:
!pip install langchain


Integrating OpenAI and Rule-Based Systems with LangChain and Gradio

This Python script implements a chatbot application using Gradio for the web interface and the Hugging Face transformers library for natural language processing. The chatbot is designed to generate responses based on predefined intents and patterns, supplemented by a GPT-2 model for more generalized responses.

First, the script loads the predefined intents from a JSON file (intents.json). These intents contain patterns and corresponding responses that guide the chatbot's behavior. The GPT-2 model and its tokenizer are then loaded from the Hugging Face model repository, and a text generation pipeline is created to facilitate the generation of responses.

The classify_intent function checks the user's query against the predefined patterns within the intents. If a match is found, it returns a random response from the corresponding intent. If no match is found, the chatbot uses the GPT-2 model to generate a response. This is handled by the get_chatbot_response function, which first attempts to find a classified response and falls back on the GPT-2 model if necessary.

The script also includes a simple rule-based system, implemented in the rule_based_system function, which suggests activities based on the user's weather and preference inputs.

The final response combines the generated text from the GPT-2 model and the activity suggestion from the rule-based system, which is then returned to the user. The Gradio interface facilitates user interaction by providing a simple form with text inputs for the query, weather, and preference, and an output box for the chatbot’s response.

Finally, the Gradio app is launched with interface.launch(), making the chatbot accessible for interaction.

In [17]:
import json
import random
import gradio as gr
from transformers import GPT2LMHeadModel, GPT2Tokenizer, pipeline

# Load the intents file
with open('/content/intents.json') as file:
    intents = json.load(file)

# Load the GPT-2 model and tokenizer
model_name = "gpt2"
tokenizer = GPT2Tokenizer.from_pretrained(model_name)
model = GPT2LMHeadModel.from_pretrained(model_name)

# Create a pipeline
chatbot_pipeline = pipeline("text-generation", model=model, tokenizer=tokenizer)

# Function to classify the user's query
def classify_intent(query):
    for intent in intents['intents']:
        for pattern in intent['patterns']:
            if pattern.lower() in query.lower():
                return random.choice(intent['responses'])
    return None

# Define a function to implement the rule-based system
def rule_based_system(weather, preference):
    if weather == "sunny" and preference == "outdoor":
        return "go for a walk"
    elif weather == "sunny" and preference == "indoor":
        return "go swimming"
    elif weather == "rainy" and preference == "outdoor":
        return "read a book"
    elif weather == "rainy" and preference == "indoor":
        return "watch a movie"
    else:
        return "stay indoors"

# Define a function to get the chatbot response using Hugging Face transformers and rule-based system
def get_chatbot_response(query, weather, preference):
    try:
        # Find the matching intent
        classified_response = classify_intent(query)
        if classified_response:
            return classified_response

        # Generate a response using Hugging Face pipeline
        response = chatbot_pipeline(
            query,
            max_length=50,
            num_return_sequences=1,
            temperature=0.7,  # Controls randomness
            top_k=50,         # Limits to top 50 probable next words
            top_p=0.95        # Nucleus sampling, focuses on top 95% probable words
        )
        response_text = response[0]['generated_text']

        # Get rule-based system response
        rule_based_response = rule_based_system(weather, preference)

        # Combine the responses
        final_response = f"{response_text}\n\nSuggested activity: {rule_based_response}"

        return final_response
    except Exception as e:
        return str(e)

# Create a Gradio interface
interface = gr.Interface(
    fn=get_chatbot_response,
    inputs=[
        gr.Textbox(label="Query"),
        gr.Dropdown(["sunny", "rainy"], label="Weather"),
        gr.Dropdown(["outdoor", "indoor"], label="Preference")
    ],
    outputs=gr.Textbox(),
    title="Chatbot with Logical Inference",
    description="A simple chatbot using Hugging Face GPT-2 model with predefined intents and logical inference."
)

# Launch the Gradio app
interface.launch()


/usr/local/lib/python3.10/dist-packages/gradio/analytics.py:106: UserWarning: IMPORTANT: You are using gradio version 3.50.2, however version 4.29.0 is available, please upgrade. 
--------
  


Setting queue=True in a Colab notebook requires sharing enabled. Setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on public URL: https://244de6cbacfc9f0536.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)


Testing and Refinement

STEP 4

Testing the Chatbot with Various Weather and Preference Conditions

This script is designed to simulate and test the responses of a chatbot that integrates rule-based logic for recommending activities based on weather and user preferences. The core function, rule_based_system, employs the python-constraint library to define and solve a constraint satisfaction problem. It considers possible activities and applies constraints based on the given weather conditions and user preferences. For example, if the weather is sunny and the preference is outdoor, suitable activities might include going for a walk or swimming.

To facilitate testing without consuming API quotas, the mock_get_chatbot_response function simulates the chatbot's behavior by directly calling the rule_based_system function, bypassing the actual OpenAI API call. This function returns a response based purely on the rule-based logic.

The test_cases list outlines various scenarios with different combinations of queries, weather conditions, and user preferences. The test_mock_chatbot function iterates through these scenarios, invoking the mock response function and printing the results. This function provides a structured way to validate the chatbot's responses to ensure they align with the expected logic without incurring any API usage.

Overall, this setup allows for comprehensive testing of the chatbot's decision-making process in a controlled environment, focusing on the logic and consistency of the responses generated by the rule-based system.

In [18]:
# Mock function to simulate get_chatbot_response without calling the API
def mock_get_chatbot_response(query, weather, preference):
    # Directly get rule-based system response without OpenAI mock text
    rule_based_response = rule_based_system(weather, preference)
    return rule_based_response

# Define test cases
test_cases = [
    {"query": "What should I do?", "weather": "sunny", "preference": "outdoor"},
    {"query": "What should I do?", "weather": "sunny", "preference": "indoor"},
    {"query": "What should I do?", "weather": "rainy", "preference": "outdoor"},
    {"query": "What should I do?", "weather": "rainy", "preference": "indoor"},
    {"query": "What should I do?", "weather": "stormy", "preference": "outdoor"}  # Edge case
]

# Function to test the chatbot with mocked responses
def test_mock_chatbot(test_cases):
    for i, case in enumerate(test_cases):
        print(f"Test Case {i+1}:")
        query = case["query"]
        weather = case["weather"]
        preference = case["preference"]
        print(f"Query: {query}")
        print(f"Weather: {weather}")
        print(f"Preference: {preference}")
        response = mock_get_chatbot_response(query, weather, preference)
        print(f"Response: {response}")
        print("-" * 50)

# Run the mock tests
test_mock_chatbot(test_cases)


Test Case 1:
Query: What should I do?
Weather: sunny
Preference: outdoor
Response: go for a walk
--------------------------------------------------
Test Case 2:
Query: What should I do?
Weather: sunny
Preference: indoor
Response: go swimming
--------------------------------------------------
Test Case 3:
Query: What should I do?
Weather: rainy
Preference: outdoor
Response: read a book
--------------------------------------------------
Test Case 4:
Query: What should I do?
Weather: rainy
Preference: indoor
Response: watch a movie
--------------------------------------------------
Test Case 5:
Query: What should I do?
Weather: stormy
Preference: outdoor
Response: stay indoors
--------------------------------------------------


The output presented is a series of test cases for a rule-based chatbot designed to recommend activities based on weather conditions and user preferences. Each test case outlines the specific query, weather condition, and user preference provided as inputs. The corresponding response demonstrates the chatbot's ability to suggest appropriate activities that align with the given conditions.

For Test Case 1, with the query "What should I do?", sunny weather, and an outdoor preference, the chatbot suggests "go for a walk." In Test Case 2, under the same weather conditions but with an indoor preference, the chatbot recommends "go swimming." For Test Case 3, with rainy weather and an outdoor preference, the response is "read a book," acknowledging the impracticality of outdoor activities in the rain. Test Case 4, with rainy weather and an indoor preference, results in the recommendation to "watch a movie," which suits the indoor preference. Lastly, Test Case 5, with stormy weather and an outdoor preference, sensibly suggests to "stay indoors," prioritizing safety.

These outputs demonstrate the chatbot's logical inference capabilities, validating that the rule-based system effectively adapts its recommendations to varying weather conditions and user preferences, ensuring appropriate and practical advice.